app.py

```diff
@@ -0,0 +1,224 @@
+    forgot = request.args.get('forgot', '')
+    stage = "1"
+    return render_template('index.html', forgot=forgot, stage=stage)
+
+@app.route('/index_action', methods=['POST'])
+def index_action():
+    forgot = request.form.get('forgot', '')
+    stage = request.form.get('stage', '')
+    email = request.form.get('email').lower()
+    answer = request.form.get('answer', '').lower()
+    query = sql_select('SELECT user_id, question, answer, username FROM users WHERE email=%s', email)
+    if query:
+        response = query[0]        
+        question = get_questions(response[1])        
+        msg = ''
+        if stage == "3":
+            if not check(answer, response[2]):                
+                stage = "2"
+                msg = 'Incorrect! Please try again...'
+        elif stage == "4":
+            user_id = query[0][0] 
+            new_password = hash(request.form.get('new_password'))
+            sql_write('UPDATE users SET password =%s WHERE user_id=%s', new_password, user_id)
+            msg = f"Password successfully updated for user '{response[3]}'"
+            return render_template('index.html', msg=msg)
+    else:
+        stage = "1"
+        question = ""
+        msg = 'Invalid email address'
+    return render_template('index.html', forgot=forgot, stage=stage, email=email, question=question, msg=msg)
+
+
+@app.route('/login_action', methods=['POST'])
+def login_action():
+    email = request.form.get('email', '').lower()
+    password = request.form.get('password', '')
+    query = sql_select('SELECT * FROM users WHERE email=%s', email)        
+    if query:
+        response = query[0]
+        if check(password, response[3]):
+            session['user_id'] = response[0]
+            session['username'] = response[1]
+            session['admin'] = response[6]
+            return redirect('/home')
+        else:
+            msg = "Incorrect username and/or password"
+        return render_template('index.html', msg=msg, query=query)
+    else:
+        msg = "Incorrect username and/or password"
+        return render_template('index.html', msg=msg, query=query)     
+
+@app.route('/signup')
+def signup():
+    questions = get_questions()
+    return render_template('signup.html', questions=questions)
+
+@app.route('/signup_action', methods=['POST'])
+def signup_action():
+    username = request.form.get('username', '').lower()
+    email = request.form.get('email', '').lower()
+    password = request.form.get('password', '')
+    verify_password = request.form.get('verify_password', '')
+    question = request.form.get('questions', '')
+    answer = request.form.get('answer', '').lower()
+    hash_answer = hash(answer)
+    # Prompt user if failed password verification  ---SWITCH THESE CHECKS TO JAVASCRIPT VALIDATION?
+    if password != verify_password:
+        msg = "Passwords do not match"
+        return render_template('signup.html', msg=msg)
+    # Check if username in use
```

Comment:
Inline commetns can often indicate a code smell and can often get of date. You can make code readable without comments using functions. E.g. this could be:

```
const isUsernameUsed = checkIfUsernameInUse()
```

a reader of this code doens't need to go into this code unless they really want to understand how it works.